src/sync/impl/sync_file.cpp

```diff
@@ -0,0 +1,316 @@
+        buffer.append(extension.substr(1));
+    } else if (path_last == '.' || extension_first == '.') {
+        buffer.append(extension);
+    } else {
+        buffer.append(".");
+        buffer.append(extension);
+    }
+    return buffer;
+}
+
+} // util
+
+constexpr const char SyncFileManager::c_sync_directory[];
+constexpr const char SyncFileManager::c_utility_directory[];
+constexpr const char SyncFileManager::c_metadata_directory[];
+constexpr const char SyncFileManager::c_metadata_realm[];
+
+std::string SyncFileManager::get_utility_directory()
+{
+    auto util_path = file_path_by_appending_component(get_base_sync_directory(),
+                                                      c_utility_directory,
+                                                      util::FilePathType::Directory);
+    util::try_make_dir(util_path);
+    return util_path;
+}
+
+std::string SyncFileManager::get_base_sync_directory()
+{
+    auto sync_path = file_path_by_appending_component(m_base_path,
+                                                      c_sync_directory,
+                                                      util::FilePathType::Directory);
+    util::try_make_dir(sync_path);
+    return sync_path;
+}
+
+std::string SyncFileManager::user_directory(const std::string& user_identity)
+{
+    REALM_ASSERT(user_identity.length() > 0);
+    if (filename_is_reserved(user_identity)) {
+        throw std::invalid_argument("A user can't have an identifier reserved by the filesystem.");
+    }
+    auto user_path = file_path_by_appending_component(get_base_sync_directory(),
+                                                      user_identity,
+                                                      util::FilePathType::Directory);
+    util::try_make_dir(user_path);
+    return user_path;
+}
+
+void SyncFileManager::remove_user_directory(const std::string& user_identity)
+{
+    REALM_ASSERT(user_identity.length() > 0);
+    if (filename_is_reserved(user_identity)) {
+        throw std::invalid_argument("A user can't have an identifier reserved by the filesystem.");
+    }
+    auto user_path = file_path_by_appending_component(get_base_sync_directory(),
+                                                      user_identity,
+                                                      util::FilePathType::Directory);
+    util::remove_nonempty_dir(user_path);
+}
+
+bool SyncFileManager::remove_realm(const std::string& user_identity, const std::string& raw_realm_path)
+{
+    REALM_ASSERT(user_identity.length() > 0);
+    REALM_ASSERT(raw_realm_path.length() > 0);
+    if (filename_is_reserved(user_identity) || filename_is_reserved(raw_realm_path)) {
+        throw std::invalid_argument("A user or Realm can't have an identifier reserved by the filesystem.");
+    }
+    auto escaped = util::make_percent_encoded_string(raw_realm_path);
+    auto realm_path = util::file_path_by_appending_component(user_directory(user_identity), std::move(escaped));
```

Comment:
No reason to move here as `file_path_by_appending_component` takes a const reference.